### DISCUSSION TITLE
Is there a way to activate auto-retry on http error response

### DISCUSSION THREAD

**Discussion by PFaurel** [encode/httpx#1677] [Q&A]
Hi,

Is there any way to rely on an auto-retry in httpx lib with http requests when response is temporary in error (4xx or 5xx).
We can do that with requests and urllib3.

Examples:
```
max_retries = urllib3.util.Retry(
            3,
            allowed_methods=frozenset({'POST'}),
            status=3,
            backoff_factor=1,
            status_forcelist=[429])

# with requests
http_session = requests.Session()
http_session.mount(
            'http://calc.com/sum',
            requests.adapters.HTTPAdapter(max_retries=max_retries)
)
result = http_session.request('GET', 'http://calc.com/sum')

# with urllib3.PoolManager
pool_mgr = urllib3.PoolManager()
result = pool_mgr.request('GET', 'http://calc2.com/sum', retries=max_retries)
```

**Comment by StephenBrown2** [comment]
Retries have been suggested a couple times: https://github.com/encode/httpx/issues/108 https://github.com/encode/httpx/pull/778 https://github.com/encode/httpx/issues/1141 https://github.com/encode/httpx/pull/1196 but I think were dropped in favor of letting users do their own thing via Event Hooks: https://github.com/encode/httpx/pull/1246 | https://www.python-httpx.org/advanced/#event-hooks and Custom Transports: https://www.python-httpx.org/advanced/#custom-transports which have an example for connection retries using the default HTTPTransport:

> Connection retries are also available via this interface.
> ```py
> >>> import httpx
> >>> transport = httpx.HTTPTransport(retries=1)
> >>> client = httpx.Client(transport=transport)
> ```

Status code retries are left as an activity for the user, however https://github.com/encode/httpx/issues/984 is still open and I think a RetryTransport would be a good example in addition to the ThrottleTransport example given in that issue by @tomchristie.
